### PR TITLE
WT-7164 Merge Restructure history store project branch to develop

### PR DESCRIPTION
**Pull request by bvpvamsikrishna**
This is the project branch from Restructure history store project.

I tried to summarise the changes done as part of the restructure history store project [here](https://docs.google.com/document/d/1NC8VdyJD7B6lUZ8ojMGK9laHJDkJU9nek3mHbQvTk7c/edit?usp=sharing). 

The implementation of history store cursor is already part of the develop and hence those changes will not appear in this PR diff.
They can be found in files [src/cursor/cur_hs.c](https://github.com/wiredtiger/wiredtiger/blob/develop/src/cursor/cur_hs.c).
 
The changes in this PR represent the changes to the users of the history store from "old way of accessing the history store" --> "using new history store cursor".

Feel free to tag me if anyone needs more context for the review.
This project branch is also under complete WT testing for the past two weeks successfully.

### CODE REVIEW

**Comment by tetsuo-cpp** [comment]
### Areas to focus on

@luke-pearson: Reconciliation from checkpoint point of view
@tetsuo-cpp: Reconciliation from eviction point of view
@kommiharibabu: Rollback to stable

File: src/cursor/cur_hs.c [hunk 1]
```diff
@@ -646,14 +721,25 @@ __curhs_set_value(WT_CURSOR *cursor, ...)
     WT_CURSOR_HS *hs_cursor;
     va_list ap;
```
Comment (luke-pearson) [inline src/cursor/cur_hs.c:723]: Unnecessary whitespace.

File: src/history/hs_rec.c [hunk 2]
```diff
@@ -207,54 +113,48 @@ __hs_insert_record_with_btree(WT_SESSION_IMPL *session, WT_CURSOR *cursor, WT_BT
      * timestamp. Otherwise the newly inserting history store record may fall behind the existing
      * one can lead to wrong order.
      */
-    WT_ERR_NOTFOUND_OK(
-      __wt_hs_cursor_position(session, cursor, btree->id, key, tw->start_ts, srch_key), true);
+    cursor->set_key(cursor, 4, btree->id, key, tw->start_ts, UINT64_MAX);
```
Comment (luke-pearson) [inline src/history/hs_rec.c:116]: 4?

File: src/history/hs_rec.c [hunk 3]
```diff
@@ -973,45 +788,39 @@ __hs_fixup_out_of_order_from_pos(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor,
...
+        insert_cursor->set_key(insert_cursor, 4, btree->id, &hs_key, ts, *counter);
+        insert_cursor->set_value(insert_cursor, &hs_insert_tw, hs_insert_tw.durable_stop_ts,
+          hs_insert_tw.durable_start_ts, (uint64_t)hs_upd_type, &hs_value);
+        WT_ERR(insert_cursor->insert(insert_cursor));
 
-        /* Reinsert entry with earlier timestamp. */
-        while ((ret = __hs_insert_record_with_btree_int(session, insert_cursor, btree->id, key,
-                  (uint8_t)hs_upd_type, &hs_value, &tw, *counter)) == WT_RESTART)
-            ;
         WT_ERR(ret);
```
Comment (kommiharibabu) [inline src/history/hs_rec.c:812]: Extra error check. Remove it.

File: src/history/hs_rec.c [hunk 4]
```diff
@@ -973,45 +788,39 @@ __hs_fixup_out_of_order_from_pos(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor,
...
-        WT_ERR(__wt_upd_alloc_tombstone(session, &tombstone, NULL));
-        tombstone->txnid = WT_TXN_NONE;
-        tombstone->start_ts = tombstone->durable_ts = WT_TS_NONE;
-        while ((ret = __wt_hs_modify(hs_cbt, tombstone)) == WT_RESTART) {
-            WT_WITH_PAGE_INDEX(session, ret = __wt_hs_row_search(hs_cbt, &hs_cursor->key, false));
-            WT_ERR(ret);
-        }
+        /* Delete the entry with higher timestamp. */
+        hs_cursor->remove(hs_cursor);
         WT_ERR(ret);
```
Comment (kommiharibabu) [inline src/history/hs_rec.c:817]: unnecessary error check

File: src/history/hs_rec.c [hunk 5]
```diff
@@ -1078,82 +882,51 @@ __hs_delete_key_from_pos(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor, uint32_
...
+            WT_STAT_CONN_INCR(session, cache_hs_insert);
+            WT_STAT_DATA_INCR(session, cache_hs_insert);
 
-            /* Reinsert entry with zero timestamp. */
-            while (
-              (ret = __hs_insert_record_with_btree_int(session, insert_cursor, btree_id, &hs_key,
-                 (uint8_t)hs_upd_type, &hs_value, &tw, hs_insert_counter)) == WT_RESTART)
-                ;
             hs_insert_counter++;
             WT_ERR(ret);
```
Comment (kommiharibabu) [inline src/history/hs_rec.c:907]: unnecessary error check.

File: dist/s_void [hunk 6]
```diff
@@ -135,7 +135,9 @@ func_ok()
 	    -e '/int zlib_terminate$/d' \
 	    -e '/int zstd_error$/d' \
 	    -e '/int zstd_pre_size$/d' \
-	    -e '/int zstd_terminate$/d'
+	    -e '/int zstd_terminate$/d' \
+        -e '/int __wt_hs_cursor_search_near_after$/d' \
```
Comment (tetsuo-cpp) [inline dist/s_void:139]: Nit: Indentation

File: src/btree/bt_debug.c [hunk 7]
```diff
@@ -1028,43 +1012,48 @@ __debug_page(WT_DBG *ds, WT_REF *ref, uint32_t flags)
...
+            WT_ERR(__debug_page_row_leaf(ds, ref->page, hs_cursor));
         break;
     default:
-        return (__wt_illegal_value(session, ref->page->type));
+        WT_ERR(__wt_illegal_value(session, ref->page->type));
     }
 
-    return (0);
+err:
+    if (hs_cursor != NULL) {
```
Comment (tetsuo-cpp) [inline src/btree/bt_debug.c:1053]: Nit: Braces

File: src/btree/bt_vrfy.c [hunk 8]
```diff
@@ -782,11 +782,12 @@ __verify_key_hs(
     wt_timestamp_t older_start_ts, older_stop_ts;
     uint64_t hs_counter;
     uint32_t hs_btree_id;
-    int cmp, exact;
     char ts_string[2][WT_TS_INT_STRING_SIZE];
 
     btree = S2BT(session);
     hs_btree_id = btree->id;
+    WT_RET(__wt_curhs_open(session, NULL, &hs_cursor));
```
Comment (tetsuo-cpp) [inline src/btree/bt_vrfy.c:789]: Do you not need to close this in this function?

File: src/cursor/cur_hs.c [hunk 9]
```diff
@@ -11,7 +11,7 @@
 
 static int __curhs_prev_visible(WT_SESSION_IMPL *, WT_CURSOR_HS *);
 static int __curhs_next_visible(WT_SESSION_IMPL *, WT_CURSOR_HS *);
-
+static int __hs_cursor_search_near_helper(WT_SESSION_IMPL *, WT_CURSOR *, bool);
```
Comment (tetsuo-cpp) [inline src/cursor/cur_hs.c:14]: What's the reason behind this naming convention compared with the `__curhs` functions? It seems like an inconsistency.

File: src/docs/Doxyfile [hunk 10]
```diff
@@ -1644,11 +1644,6 @@ ALLEXTERNALS           = NO
 
 EXTERNAL_GROUPS        = YES
 
-# The PERL_PATH should be the absolute path and name of the perl script
```
Comment (tetsuo-cpp) [inline src/docs/Doxyfile:1647]: How does this part fit into the change?

**Review by tetsuo-cpp — COMMENTED** [review]
Just some initial comments. Will take a closer look this afternoon.

File: src/cursor/cur_hs.c [hunk 11]
```diff
@@ -488,153 +598,118 @@ __curhs_search_near(WT_CURSOR *cursor, int *exactp)
...
-         * place us at the end of a particular key/timestamp range whereas we want to be placed at
-         * the beginning.
-         */
-        if (exact < 0) {
-            while ((ret = __wt_hs_cursor_next(session, file_cursor)) == 0) {
-                WT_ERR(__wt_compare(session, NULL, &file_cursor->key, srch_key, &cmp));
-                if (cmp >= 0)
-                    break;
-            }
+    if (exact >= 0) {
```
[the remark this answers is not in the document]
Comment (luke-pearson) [reply src/cursor/cur_hs.c:625]: A lot more would be good.

File: src/history/hs_cursor.c [hunk 12]
```diff
@@ -87,133 +87,66 @@ __wt_hs_modify(WT_CURSOR_BTREE *hs_cbt, WT_UPDATE *hs_upd)
...
-        WT_ASSERT(session, cmp <= 0);
-    }
-#endif
-err:
-    if (user_srch_key == NULL)
-        __wt_scr_free(session, &srch_key);
-    return (ret);
+    file_cursor = ((WT_CURSOR_HS *)hs_cursor)->file_cursor;
+    hs_cbt = (WT_CURSOR_BTREE *)file_cursor;
+    *twp = &hs_cbt->upd_value->tw;
```
[the remark this answers is not in the document]
Comment (luke-pearson) [reply src/history/hs_cursor.c:101]: That should be macro and also used in: `__wt_hs_cbt`

File: src/history/hs_rec.c [hunk 13]
```diff
@@ -207,54 +113,48 @@ __hs_insert_record_with_btree(WT_SESSION_IMPL *session, WT_CURSOR *cursor, WT_BT
...
-            }
-#else
+        if (tw->start_ts == hs_start_ts) {
+            WT_ERR(cursor->get_value(cursor, &hs_stop_durable_ts_diag, &durable_timestamp_diag,
+              &upd_type_full_diag, existing_val));
+            WT_ERR(__wt_compare(session, NULL, existing_val, hs_value, &cmp));
             if (cmp == 0)
-                counter = hs_counter + 1;
-#endif
+                WT_ASSERT(session,
```
Comment (luke-pearson) [inline src/history/hs_rec.c:128]: Please put a comment on this assert.

File: src/cursor/cur_hs.c [hunk 14]
```diff
@@ -479,6 +529,67 @@ __curhs_next_visible(WT_SESSION_IMPL *session, WT_CURSOR_HS *hs_cursor)
...
+}
+
+/*
+ * __wt_hs_cursor_search_near_after --
+ *     Set the cursor position at the requested position or after it.
+ */
+int
+__wt_hs_cursor_search_near_after(WT_SESSION_IMPL *session, WT_CURSOR *cursor)
+{
+    return __hs_cursor_search_near_helper(session, cursor, false);
```
Comment (luke-pearson) [inline src/cursor/cur_hs.c:549]: Again needs to be wrapped.

File: src/history/hs_rec.c (hunk 3 above)
Comment (hbokhari) [reply src/history/hs_rec.c:812]: Fixed.

File: src/history/hs_rec.c (hunk 4 above)
Comment (hbokhari) [reply src/history/hs_rec.c:817]: Fixed.

File: src/history/hs_rec.c (hunk 5 above)
Comment (hbokhari) [reply src/history/hs_rec.c:907]: Fixed.

File: src/cursor/cur_hs.c [hunk 15]
```diff
@@ -479,6 +528,67 @@ __curhs_next_visible(WT_SESSION_IMPL *session, WT_CURSOR_HS *hs_cursor)
...
 }
 
+/*
+ * __wt_hs_cursor_search_near_before --
+ *     Set the cursor position at the requested position or before it.
+ */
+int
+__wt_hs_cursor_search_near_before(WT_SESSION_IMPL *session, WT_CURSOR *cursor)
+{
+    return __hs_cursor_search_near_helper(session, cursor, true);
```
[the remark this answers is not in the document]
Comment (hbokhari) [reply src/cursor/cur_hs.c:539]: Fixed.

File: src/cursor/cur_hs.c (hunk 1 above)
Comment (hbokhari) [reply src/cursor/cur_hs.c:724]: Fixed.

File: src/cursor/cur_hs.c (hunk 14 above)
Comment (hbokhari) [reply src/cursor/cur_hs.c:549]: Fixed.

File: src/btree/bt_debug.c (hunk 7 above)
Comment (hbokhari) [reply src/btree/bt_debug.c:1053]: Fixed.

File: src/history/hs_rec.c (hunk 13 above)
Comment (hbokhari) [reply src/history/hs_rec.c:128]: Added a comment to explain the assert.

File: src/history/hs_cursor.c [hunk 16]
```diff
@@ -87,133 +87,66 @@ __wt_hs_modify(WT_CURSOR_BTREE *hs_cbt, WT_UPDATE *hs_upd)
 }
 
 /*
- * __hs_cursor_position_int --
- *     Internal function to position a history store cursor at the end of a set of updates for a
- *     given btree id, record key and timestamp.
+ * __wt_hs_upd_time_window --
```
[the remark this answers is not in the document]
Comment (hbokhari) [reply src/history/hs_cursor.c:90]: Fixed.

File: src/btree/bt_vrfy.c (hunk 8 above)
Comment (hbokhari) [reply src/btree/bt_vrfy.c:785]: Nice catch - yes we need to close the cursor. Fixed.

File: dist/s_void (hunk 6 above)
Comment (hbokhari) [reply dist/s_void]: Fixed.

File: src/cursor/cur_hs.c [hunk 17]
```diff
@@ -145,6 +145,32 @@ __wt_hs_cursor_search_near(WT_SESSION_IMPL *session, WT_CURSOR *cursor, int *exa
     return (ret);
 }
 
+/*
+ * __curhs_key_return --
+ *     return hs key
+ */
+static inline void
+__curhs_key_return(WT_CURSOR *hs_cursor, WT_CURSOR *file_cursor)
```
[the remark this answers is not in the document]
Comment (hbokhari) [reply src/cursor/cur_hs.c]: Fixed.

File: src/cursor/cur_hs.c [hunk 18]
```diff
@@ -145,6 +145,32 @@ __wt_hs_cursor_search_near(WT_SESSION_IMPL *session, WT_CURSOR *cursor, int *exa
...
+{
+    hs_cursor->key.data = file_cursor->key.data;
+    hs_cursor->key.size = file_cursor->key.size;
+    WT_ASSERT(CUR2S(file_cursor), F_ISSET(file_cursor, WT_CURSTD_KEY_SET));
+    F_SET(hs_cursor, F_MASK(file_cursor, WT_CURSTD_KEY_SET));
+}
+
+/*
+ * __curhs_value_return --
+ *     return hs value
```
[the remark this answers is not in the document]
Comment (hbokhari) [reply src/cursor/cur_hs.c]: Fixed.

File: src/cursor/cur_hs.c [hunk 19]
```diff
@@ -380,6 +419,11 @@ __curhs_prev_visible(WT_SESSION_IMPL *session, WT_CURSOR_HS *hs_cursor)
         if (F_ISSET(std_cursor, WT_CURSTD_HS_READ_COMMITTED))
             break;
 
+        /*
+         * If we don't have a transaction snapshot, we should have already set the appropriate flag.
```
[the remark this answers is not in the document]
Comment (hbokhari) [reply src/cursor/cur_hs.c]: I have changed the comment. Let me know if you think it needs further clarification.

File: src/docs/Doxyfile (hunk 10 above)
Comment (hbokhari) [reply src/docs/Doxyfile]: Not sure how this change got here. I have reverted it.

File: src/history/hs_cursor.c (hunk 12 above)
[the remark this answers is not in the document]
Comment (hbokhari) [reply src/history/hs_cursor.c:99]: Fixed. We have now inlined version of `__wt_hs_cbt`

File: src/include/cursor_inline.h [hunk 20]
```diff
@@ -6,6 +6,32 @@
...
+ * __wt_curhs_btree --
+ *     Convert a history store cursor to the underlying btree.
+ */
+static inline WT_BTREE *
+__wt_curhs_btree(WT_CURSOR *cursor)
+{
+    WT_CURSOR_HS *hs_cursor;
+    hs_cursor = (WT_CURSOR_HS *)cursor;
+
+    return CUR2BT(hs_cursor->file_cursor);
```
Comment (luke-pearson) [inline src/include/cursor_inline.h:19]: Could you return `(WT_BTREE *)__wt_curhs_cbt`?

File: src/history/hs_cursor.c [hunk 21]
```diff
@@ -440,26 +293,8 @@ __hs_find_upd_int(WT_SESSION_IMPL *session, uint32_t btree_id, WT_ITEM *key,
...
-    WT_BTREE *btree;
-    WT_DECL_RET;
-
-    btree = S2BT(session);
+    if (hs_cursor != NULL)
+        WT_TRET(hs_cursor->close(hs_cursor));
 
-    WT_RET(__wt_hs_cursor_open(session));
-    WT_WITH_BTREE(session, CUR2BT(session->hs_cursor),
-      (ret = __hs_find_upd_int(
```
[the remark this answers is not in the document]
Comment (bvpvamsikrishna) [reply src/history/hs_cursor.c:461]: Yes, now we have implemented the history store cursor.  This means all the history store cursor functions/apis will call "CURSOR_API_CALL_PREPARE_ALLOWED", which will take care of saving the session state and restoring it back, which includes setting the btree in session to history store and restoring the original btree at the end of the function.

File: src/include/cursor_inline.h (hunk 20 above)
Comment (hbokhari) [reply src/include/cursor_inline.h:19]: Nope. `btree` is embedded deeper. Have a look at `CUR2BT()` macro.

File: src/cursor/cur_hs.c [hunk 22]
```diff
@@ -479,6 +528,67 @@ __curhs_next_visible(WT_SESSION_IMPL *session, WT_CURSOR_HS *hs_cursor)
...
+__hs_cursor_search_near_helper(WT_SESSION_IMPL *session, WT_CURSOR *cursor, bool before)
+{
+    WT_DECL_ITEM(srch_key);
+    WT_DECL_RET;
+    int cmp;
+
+    WT_ERR(__wt_scr_alloc(session, 0, &srch_key));
+    WT_ERR(__wt_buf_set(session, srch_key, cursor->key.data, cursor->key.size));
+    WT_ERR(cursor->search_near(cursor, &cmp));
+    if (before) {
```
[the remark this answers is not in the document]
Comment (hbokhari) [reply src/cursor/cur_hs.c:540]: We have added comments.

File: src/cursor/cur_hs.c (hunk 11 above)
[the remark this answers is not in the document]
Comment (hbokhari) [reply src/cursor/cur_hs.c:609]: We have added comments. Let me know if you think we need to add more comments here.

File: src/history/hs_rec.c (hunk 2 above)
Comment (hbokhari) [reply src/history/hs_rec.c:113]: It's the number of arguments after the first fixed cursor argument.

File: src/cursor/cur_hs.c (hunk 9 above)
Comment (hbokhari) [reply src/cursor/cur_hs.c]: We have renamed a number of functions in cur_hs.c file to make them following the correct naming conventions.

File: src/btree/bt_vrfy.c [hunk 23]
```diff
@@ -833,7 +819,7 @@ __verify_key_hs(
          */
         newer_start_ts = older_start_ts;
     }
-
+    WT_TRET(hs_cursor->close(hs_cursor));
```
Comment (luke-pearson) [inline src/btree/bt_vrfy.c:822]: Shouldn't this go in an err: block so if anything else returns a failure we still close the cursor?

File: src/cursor/cur_hs.c [hunk 24]
```diff
@@ -380,6 +396,12 @@ __curhs_prev_visible(WT_SESSION_IMPL *session, WT_CURSOR_HS *hs_cursor)
         if (F_ISSET(std_cursor, WT_CURSTD_HS_READ_COMMITTED))
             break;
 
+        /*
+         * If we are using history store cursor and do not have a snapshot, we should set the flag
```
Comment (luke-pearson) [inline src/cursor/cur_hs.c:400]: This comment is still not great, perhaps something like:

"If we are using a history store cursor and haven't set the WT_CURSTD_HS_READ_COMMITTED flag then we must have a snapshot, assert that we do."

Comment (luke-pearson) [reply src/cursor/cur_hs.c:400]: Again the same below.

File: src/cursor/cur_hs.c [hunk 25]
```diff
@@ -479,186 +507,287 @@ __curhs_next_visible(WT_SESSION_IMPL *session, WT_CURSOR_HS *hs_cursor)
...
+        /*
+         * If we want to land on a key that is smaller or equal to the specified key, keep walking
+         * backwards as there may be content inserted concurrently.
+         */
+        if (cmp > 0) {
+            while ((ret = cursor->prev(cursor)) == 0) {
+                WT_STAT_CONN_INCR(session, cursor_skip_hs_cur_position);
+                WT_STAT_DATA_INCR(session, cursor_skip_hs_cur_position);
+                WT_ERR(__wt_compare(session, NULL, &cursor->key, srch_key, &cmp));
+                /* We find a key that is smaller or equal to the specified key. */
```
Comment (luke-pearson) [inline src/cursor/cur_hs.c:554]: This comment wording can be improved to something like: "Have we found a key that is smaller or equal..."

Comment (luke-pearson) [reply src/cursor/cur_hs.c:554]: Same below.

File: src/cursor/cur_hs.c (hunk 24 above)
Comment (hbokhari) [reply src/cursor/cur_hs.c]: Fixed.

File: src/cursor/cur_hs.c (hunk 25 above)
Comment (hbokhari) [reply src/cursor/cur_hs.c]: Fixed.

File: src/btree/bt_vrfy.c (hunk 23 above)
Comment (hbokhari) [reply src/btree/bt_vrfy.c]: Fixed in 5478b37.

**Review by luke-pearson — APPROVED** [review]
LGTM.

File: src/btree/bt_vrfy.c [hunk 26]
```diff
@@ -795,36 +794,23 @@ __verify_key_hs(
...
-        WT_RET(hs_cursor->get_key(hs_cursor, &hs_btree_id, vs->tmp2, &older_start_ts, &hs_counter));
-
-        if (hs_btree_id != btree->id)
-            break;
-
-        WT_RET(__wt_compare(session, NULL, tmp1, vs->tmp2, &cmp));
-        if (cmp != 0)
-            break;
-
+        WT_ERR(hs_cursor->get_key(hs_cursor, &hs_btree_id, vs->tmp2, &older_start_ts, &hs_counter));
```
Comment (tetsuo-cpp) [inline src/btree/bt_vrfy.c:801]: Why don't you need to supply the arg count here?

File: src/cursor/cur_hs.c [hunk 27]
```diff
@@ -145,9 +121,35 @@ __wt_hs_cursor_search_near(WT_SESSION_IMPL *session, WT_CURSOR *cursor, int *exa
     return (ret);
 }
 
+/*
+ * __curhs_copy_key --
+ *     Copy the key from file cursor to the history store cursor.
+ */
+static inline void
+__curhs_copy_key(WT_CURSOR *hs_cursor, WT_CURSOR *file_cursor)
```
Comment (tetsuo-cpp) [inline src/cursor/cur_hs.c:129]: I guess this is subjective, but I think the name is misleading. In my mind, copy usually means deep copy which makes it sound like it's memcpying the key buffer somewhere.

File: src/cursor/cur_hs.c [hunk 28]
```diff
@@ -479,186 +507,289 @@ __curhs_next_visible(WT_SESSION_IMPL *session, WT_CURSOR_HS *hs_cursor)
...
+         * If we want to land on a key that is smaller or equal to the specified key, keep walking
+         * backwards as there may be content inserted concurrently.
+         */
+        if (cmp > 0) {
+            while ((ret = cursor->prev(cursor)) == 0) {
+                WT_STAT_CONN_INCR(session, cursor_skip_hs_cur_position);
+                WT_STAT_DATA_INCR(session, cursor_skip_hs_cur_position);
+                WT_ERR(__wt_compare(session, NULL, &cursor->key, srch_key, &cmp));
+                /*
+                 * Exit if we have found a key that is smaller than or equal to the specified key.
```
Comment (tetsuo-cpp) [inline src/cursor/cur_hs.c:555]: Inconsistency with the comment below. May as well just do it like this:
```
/* blah */
```

File: src/history/hs_conn.c [hunk 29]
```diff
@@ -60,17 +60,16 @@ __hs_cleanup_las(WT_SESSION_IMPL *session)
 int
 __wt_hs_get_btree(WT_SESSION_IMPL *session, WT_BTREE **hs_btreep)
 {
+    WT_CURSOR *hs_cursor;
+    WT_CURSOR_BTREE *cbt;
     WT_DECL_RET;
-
```
Comment (tetsuo-cpp) [inline src/history/hs_conn.c:64]: Whitespace.

File: src/include/cursor_inline.h [hunk 30]
```diff
@@ -6,6 +6,32 @@
...
+ * __wt_curhs_cbt --
+ *     Convert a history store cursor to the underlying btree cursor.
+ */
+static inline WT_CURSOR_BTREE *
+__wt_curhs_cbt(WT_CURSOR *cursor)
+{
+    WT_CURSOR_HS *hs_cursor;
+    hs_cursor = (WT_CURSOR_HS *)cursor;
+
+    return (WT_CURSOR_BTREE *)hs_cursor->file_cursor;
```
Comment (tetsuo-cpp) [inline src/include/cursor_inline.h:32]: Formatting: Brackets around returns.

File: src/cursor/cur_hs.c [hunk 31]
```diff
@@ -253,17 +260,23 @@ __curhs_reset(WT_CURSOR *cursor)
...
     API_END_RET(session, ret);
 }
 
 /*
  * __curhs_set_key --
- *     WT_CURSOR->set_key method for the hs cursor type.
+ *     WT_CURSOR->set_key method for the history store cursor type.
  */
 static void
 __curhs_set_key(WT_CURSOR *cursor, ...)
```
Comment (tetsuo-cpp) [inline src/cursor/cur_hs.c:282]: Where's the code where you're using the explicitly passed-in argument count? I was interested in reading that part but couldn't find it.

File: src/cursor/cur_hs.c (hunk 27 above)
Comment (hbokhari) [reply src/cursor/cur_hs.c:129]: Only improvement I can think of is `__curhs_copy_key_ptr'.

File: src/cursor/cur_hs.c [hunk 32]
```diff
@@ -283,6 +296,7 @@ __curhs_set_key(WT_CURSOR *cursor, ...)
     start_ts = WT_TS_NONE;
     counter = 0;
 
+    hs_cursor->flags = 0;
     va_start(ap, cursor);
     arg_count = va_arg(ap, uint32_t);
```
Comment (hbokhari) [inline src/cursor/cur_hs.c:302]: @tetsuo-cpp - This is the code where we are using argument count.

Comment (tetsuo-cpp) [reply src/cursor/cur_hs.c:302]: Oh, I see. I didn't realise that part is already in `develop`.

File: src/cursor/cur_hs.c (hunk 28 above)
Comment (hbokhari) [reply src/cursor/cur_hs.c:555]: Sorry I didn't understand your comment.

File: src/include/cursor_inline.h (hunk 30 above)
Comment (hbokhari) [reply src/include/cursor_inline.h]: Fixed in 672dd09.

File: src/history/hs_conn.c (hunk 29 above)
Comment (hbokhari) [reply src/history/hs_conn.c:64]: Fixed in 60f2fd0.

File: src/cursor/cur_hs.c (hunk 31 above)
Comment (hbokhari) [reply src/cursor/cur_hs.c:282]: I have tagged you to the correct code line.

File: src/cursor/cur_hs.c (hunk 28 above)
Comment (luke-pearson) [reply src/cursor/cur_hs.c:555]: He's saying that the same comment below is between the comment out indicators. Whereas this one is spread over 3 lines. 

`/* blah */` vs 
```
/*
 * blah
 */
```
I think this comment may be indented one layer deeper so it may not be possible to fit it on a single line.

Comment (hbokhari) [reply src/cursor/cur_hs.c:555]: Ah okay. The character count in sentences are not equal - this one is 1 character longer and makes it overflow the 100 character limit.

Let me check if I can make it consistent any other way.

Comment (tetsuo-cpp) [reply src/cursor/cur_hs.c:555]: No worries. I thought it was just an accident.

If the comment formatting script made it like that, no need to change anything. 👍

File: src/cursor/cur_hs.c (hunk 27 above)
Comment (hbokhari) [reply src/cursor/cur_hs.c]: Fixed in 1e1d5ed.

File: src/cursor/cur_hs.c [hunk 33]
```diff
@@ -479,186 +507,289 @@ __curhs_next_visible(WT_SESSION_IMPL *session, WT_CURSOR_HS *hs_cursor)
...
+ *     Helper function to set the cursor position based on search criteria.
+ */
+static int
+__curhs_search_near_helper(WT_SESSION_IMPL *session, WT_CURSOR *cursor, bool before)
+{
+    WT_DECL_ITEM(srch_key);
+    WT_DECL_RET;
+    int cmp;
+
+    WT_ERR(__wt_scr_alloc(session, 0, &srch_key));
```
Comment (kommiharibabu) [inline src/cursor/cur_hs.c:541]: Use WT_RET.

File: src/history/hs_conn.c [hunk 34]
```diff
@@ -60,17 +60,17 @@ __hs_cleanup_las(WT_SESSION_IMPL *session)
...
+    WT_CURSOR_BTREE *cbt;
     WT_DECL_RET;
 
     *hs_btreep = NULL;
 
-    WT_RET(__wt_hs_cursor_open(session));
-
-    *hs_btreep = CUR2BT(session->hs_cursor);
+    WT_RET(__wt_curhs_open(session, NULL, &hs_cursor));
+    cbt = __wt_curhs_cbt(hs_cursor);
```
Comment (kommiharibabu) [inline src/history/hs_conn.c:70]: use `__wt_curhs_btree`

File: src/history/hs_rec.c [hunk 35]
```diff
@@ -207,54 +110,53 @@ __hs_insert_record_with_btree(WT_SESSION_IMPL *session, WT_CURSOR *cursor, WT_BT
...
 
     /*
      * If we're inserting a non-zero timestamp, look ahead for any higher timestamps. If we find
      * updates, we should remove them and reinsert them at the current timestamp.
      */
     if (tw->start_ts != WT_TS_NONE) {
-        WT_ERR_NOTFOUND_OK(__wt_hs_cursor_next(session, cursor), true);
+        /*
+         * If there were no keys equal to or less than our target key, we would have received
+         * WT_NOT_FOUND. In that case we need to search again with a higher timestamp.
```
Comment (kommiharibabu) [inline src/history/hs_rec.c:149]: WT_NOTFOUND?

Comment (kommiharibabu) [reply src/history/hs_rec.c:149]: Add more details to the comments why we are again doing the search.

File: src/history/hs_verify.c [hunk 36]
```diff
@@ -46,7 +43,7 @@ __hs_verify_id(WT_SESSION_IMPL *session, WT_CURSOR_BTREE *ds_cbt, uint32_t this_
      * verify. When we return after moving to a new key the caller is responsible for keeping the
      * cursor there or deciding they're done.
      */
-    for (; ret == 0; ret = __wt_hs_cursor_next(session, hs_cursor)) {
+    for (; ret == 0; ret = hs_cursor->next(hs_cursor)) {
         WT_ERR(hs_cursor->get_key(hs_cursor, &btree_id, &key, &hs_start_ts, &hs_counter));
 
         /*
```
Comment (kommiharibabu) [inline src/history/hs_verify.c:49]: Change the if check to assert.

**Review by kommiharibabu — APPROVED** [review]
LGTM, provided the comments are fixed.

File: src/btree/bt_debug.c [hunk 37]
```diff
@@ -970,19 +949,19 @@ __wt_debug_cursor_page(void *cursor_arg, const char *ofile)
...
     WT_DECL_RET;
     WT_SESSION_IMPL *session;
 
-    session = CUR2S(cursor_arg);
-
-    WT_RET(__wt_hs_cursor_open(session));
-    cbt = (WT_CURSOR_BTREE *)session->hs_cursor;
+    session = (WT_SESSION_IMPL *)session_arg;
+    WT_RET(__wt_curhs_open(session, NULL, &hs_cursor));
+    cbt = __wt_curhs_cbt(hs_cursor);
```
[the remark this answers is not in the document]
Comment (hbokhari) [reply src/btree/bt_debug.c:962]: Fixed.

File: src/btree/bt_debug.c [hunk 38]
```diff
@@ -1028,43 +1009,47 @@ __debug_page(WT_DBG *ds, WT_REF *ref, uint32_t flags)
      * doesn't work, we may be running in-memory.
      */
     if (!WT_IS_HS(session->dhandle)) {
-        if (session->hs_cursor != NULL || __wt_hs_cursor_open(session) == 0) {
-            WT_RET(__wt_scr_alloc(session, 0, &ds->hs_key));
-            WT_RET(__wt_scr_alloc(session, 0, &ds->hs_value));
+        if (__wt_curhs_open(session, NULL, &hs_cursor) == 0) {
```
[the remark this answers is not in the document]
Comment (hbokhari) [reply src/btree/bt_debug.c:1012]: Fixed.

File: src/btree/bt_vrfy.c [hunk 39]
```diff
@@ -795,36 +794,23 @@ __verify_key_hs(
...
      * and iterate backwards until we reach a different key or btree.
      */
-    hs_cursor = session->hs_cursor;
-    hs_cursor->set_key(hs_cursor, hs_btree_id, tmp1, WT_TS_MAX, WT_TXN_MAX);
-    ret = hs_cursor->search_near(hs_cursor, &exact);
-
-    /* If we jumped to the next key, go back to the previous key. */
-    if (ret == 0 && exact > 0)
-        ret = hs_cursor->prev(hs_cursor);
+    hs_cursor->set_key(hs_cursor, 4, hs_btree_id, tmp1, WT_TS_MAX, WT_TXN_MAX);
```
[the remark this answers is not in the document]
Comment (hbokhari) [reply src/btree/bt_vrfy.c:797]: Fixed.

File: src/history/hs_conn.c (hunk 34 above)
Comment (hbokhari) [reply src/history/hs_conn.c:70]: Fixed.

File: src/cursor/cur_hs.c (hunk 33 above)
Comment (hbokhari) [reply src/cursor/cur_hs.c:541]: Fixed.

File: src/history/hs_verify.c (hunk 36 above)
Comment (hbokhari) [reply src/history/hs_verify.c:49]: Fixed.

File: src/history/hs_rec.c (hunk 35 above)
Comment (hbokhari) [reply src/history/hs_rec.c:149]: Fixed.

File: src/cursor/cur_hs.c [hunk 40]
```diff
@@ -253,17 +260,23 @@ __curhs_reset(WT_CURSOR *cursor)
     ret = file_cursor->reset(file_cursor);
     WT_TIME_WINDOW_INIT(&hs_cursor->time_window);
     hs_cursor->btree_id = 0;
-    hs_cursor->datastore_key.data = NULL;
-    hs_cursor->datastore_key.size = 0;
+    hs_cursor->datastore_key->data = NULL;
+    hs_cursor->datastore_key->size = 0;
     hs_cursor->flags = 0;
+    cursor->key.data = cursor->key.mem;
```
[the remark this answers is not in the document]
Comment (hbokhari) [reply src/cursor/cur_hs.c:266]: Fixed

File: src/cursor/cur_hs.c [hunk 41]
```diff
@@ -145,9 +121,35 @@ __wt_hs_cursor_search_near(WT_SESSION_IMPL *session, WT_CURSOR *cursor, int *exa
     return (ret);
 }
 
+/*
+ * __curhs_copy_key_ptr --
+ *     Copy the key buffer pointer from file cursor to the history store cursor.
+ */
+static inline void
+__curhs_copy_key_ptr(WT_CURSOR *hs_cursor, WT_CURSOR *file_cursor)
```
[the remark this answers is not in the document]
Comment (hbokhari) [reply src/cursor/cur_hs.c:129]: Fixed. Changed to `__curhs_set_key_ptr` and `__curhs_set_value_ptr`.

File: src/include/cursor_inline.h [hunk 42]
```diff
@@ -6,6 +6,32 @@
  * See the file LICENSE for redistribution information.
  */
 
+/*
+ * __wt_curhs_btree --
+ *     Convert a history store cursor to the underlying btree.
+ */
+static inline WT_BTREE *
+__wt_curhs_btree(WT_CURSOR *cursor)
```
[the remark this answers is not in the document]
Comment (hbokhari) [reply src/include/cursor_inline.h]: I have renamed functions to `__wt_curhs_get_btree` and `__wt_curhs_get_cbt`.
